x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/MergedGeoLines.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.spatial.search.aggregations;
+
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.util.List;
+
+/**
+ * Class to merge an arbitrary list of {@link InternalGeoLine} lines into a new line
+ * with the appropriate max length. The final point and sort values can be found in
+ * finalPoints and finalSortValues after merge is called.
+ */
+final class MergedGeoLines {
+
+    private final List<InternalGeoLine> geoLines;
+    private final int capacity;
+    private final SortOrder sortOrder;
+    private final int[] lineIndices; // index of which geoLine item represents
+    private final int[] idxsWithinLine; // index within the geoLine for the item
+    private int size;
+    private final long[] finalPoints;       // the final sorted list of points, sorted by their respective sort-values. valid after merge
+    private final double[] finalSortValues; // the final sorted list of sort-values. valid after merge.
+
+    MergedGeoLines(List<InternalGeoLine> geoLines, int finalLength, SortOrder sortOrder) {
+        this.geoLines = geoLines;
+        this.capacity = geoLines.size();
+        this.sortOrder = sortOrder;
+        this.lineIndices = new int[capacity];
+        this.idxsWithinLine = new int[capacity];
+        this.size = 0;
+        this.finalPoints = new long[finalLength];
+        this.finalSortValues = new double[finalLength];
+    }
+
+    public long[] getFinalPoints() {
+        return finalPoints;
+    }
+
+    public double[] getFinalSortValues() {
+        return finalSortValues;
+    }
+
+    /**
+     * merges <code>geoLines</code> into one sorted list of values representing the combined line.
+     */
+    public void merge() {
+        // 1. add first element of each sub line to heap
+        for (int i = 0; i < geoLines.size(); i++) {
+            add(i, 0);
```

Comment:
What happen here if one of the geoLines has size == 0?

Reply:
good point. I will verify this. I was operating under the assumption that a new `InternalAggregation` is only made if the bucket is not empty, but that may not be the case for all parent bucketing aggregations